### PR TITLE
Feature/learner updates

### PR DESCRIPTION
**Pull request by patbaa**
Two changes:
1.) The `mli_get_current_weights` returns all trainable and non-trainable parameters. Previously only the trainable parameters were returned for PyTochLearner, but normalization layers have non-trainable parameters, which are needed too to reproduce the model performance. See running statistics [here](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html).
2.) Erase optimizer history at the beginning of the `mli_propose_weights`, which is more general, as history gets outdated when others improve the model.

The learner test files were also modified, accordingly.

### CODE REVIEW

File: colearn_keras/keras_learner.py [hunk 1]
```diff
@@ -76,13 +76,46 @@ def __init__(self, model: keras.Model,
 
         self.vote_score: float = self.test(self.train_loader)
 
+    def recompile_model(self):
+        """
+        Recompiles the Keras model. This way the optimizer history get erased,
+        which is needed before a new training round, otherwise the outdated history is used.
+        """
+        compile_args = self.model._get_compile_args()  # pylint: disable=protected-access
```
Comment (patbaa) [inline colearn_keras/keras_learner.py:84]: I did not find any better way of resetting the optimizer, than re-compiling the mode, which needs calling [this protected function](https://github.com/tensorflow/tensorflow/blob/ec00b562a4e64864cbb4311b3d9ffd97ea4bd22e/tensorflow/python/keras/engine/training.py#L2768).

When I simply replace the optimizer
```python
new_opt = keras.optimizers.SGD(lr=0.1)
model.optimizer = new_opt
```
then I cannot change batch size anymore, and also the optimizer seems not to be updated, the `model.optimizer.get_weights()` should store the number of steps and the momentums, but it remains an empty list this way.

Comment (qati) [reply colearn_keras/keras_learner.py:84]: maybe name the function `reset_optimizer` to better reflect why we have this and when we want to use it

File: colearn_keras/keras_learner.py [hunk 2]
```diff
@@ -76,13 +76,46 @@ def __init__(self, model: keras.Model,
...
+        opt_config = self.model.optimizer.get_config()
+
+        if opt_config['name'] == 'Adadelta':
+            compile_args['optimizer'] = keras.optimizers.Adadelta.from_config(opt_config)
+        elif opt_config['name'] == 'Adagrad':
```
Comment (qati) [inline colearn_keras/keras_learner.py:89]: can we maybe have an automatic version of this? As optimizers can change or extended. someting like `gettatr(keras.optimizes, opt_config['name']).from_config(...`

File: colearn_keras/keras_learner.py [hunk 3]
```diff
@@ -76,13 +76,46 @@ def __init__(self, model: keras.Model,
...
+
     def mli_propose_weights(self) -> Weights:
         """
         Trains model on training set and returns new weights after training
         - Current model is reverted to original state after training
         :return: Weights after training
         """
         current_weights = self.mli_get_current_weights()
+        # erase the outdated optimizer memory (momentums mostly)
+        self.recompile_model()
```
Comment (qati) [inline colearn_keras/keras_learner.py:118]: we always want reset? should it be configurable?

File: colearn_keras/test_keras_learner.py [hunk 4]
```diff
@@ -28,6 +28,8 @@ def get_mock_model() -> mock.Mock:
     model.evaluate.return_value = {"loss": 1,
                                    "accuracy": 3}
     model.get_weights.return_value = "all the weights"
+    model.optimizer.get_config.return_value = {"name": "Adam"}
```
Comment (qati) [inline colearn_keras/test_keras_learner.py:31]: why we need this here? do this PR changes the overall 'how to set up learner' pipeline?

File: colearn_pytorch/pytorch_learner.py [hunk 5]
```diff
@@ -90,16 +95,16 @@ def set_weights(self, weights: Weights):
...
-                old_param.set_(new_param)
+        self.model.load_state_dict(weights.weights)
 
     def train(self):
         """
         Trains the model on the training dataset
         """
 
+        # erase the outdated optimizer memory (momentums mostly)
+        self.optimizer.__setstate__({'state': defaultdict(dict)})
```
Comment (qati) [inline colearn_pytorch/pytorch_learner.py:106]: why are we reseting here? and not in the same function as in tensorflow case?

File: colearn_pytorch/test_pytorch_learner.py [hunk 6]
```diff
@@ -26,8 +27,8 @@
 from colearn_pytorch.pytorch_learner import PytorchLearner
 
 # torch does not correctly type-hint its tensor class so pylint fails
-MODEL_PARAMETERS = [torch.tensor([3, 3]), torch.tensor([4, 4])]  # pylint: disable=not-callable
-MODEL_PARAMETERS2 = [torch.tensor([5, 5]), torch.tensor([6, 6])]  # pylint: disable=not-callable
+MODEL_PARAMETERS = OrderedDict({'param1': torch.tensor([3, 3]), 'param2': torch.tensor([4, 4])})  # pylint: disable=not-callable
```
Comment (qati) [inline colearn_pytorch/test_pytorch_learner.py:30]: why we need this? does this mean we have to modify model parameters structure when we use keras learner?

File: colearn_keras/keras_learner.py (hunk 2 above)
Comment (patbaa) [reply colearn_keras/keras_learner.py:89]: I agree with both comments, fixing them...

File: colearn_keras/keras_learner.py (hunk 3 above)
Comment (patbaa) [reply colearn_keras/keras_learner.py:118]: I'd say almost always. I add a binary option need_reset_optimizer to both learners, which will have a default value of True. 

This way, all demos can remain untouched, yet still there is a way to disable optimizer resetting.

File: colearn_pytorch/pytorch_learner.py (hunk 5 above)
Comment (patbaa) [reply colearn_pytorch/pytorch_learner.py:106]: added a separate function & option to reset for the pytorchlearner too. 

We can call the reset either in the `mli_propose_weights` function right before the `self.train()` or in the beginning of the `self.train()`. There is no difference between the two. 

For both Keras and PyTorch I call the reset_optimizer in the beginning of the `self.train()` now in the fix.

File: colearn_keras/test_keras_learner.py (hunk 4 above)
Comment (patbaa) [reply colearn_keras/test_keras_learner.py:31]: No, that is not changed (the demos are still running without any modification). 

The `model` had an `optimizer` attribute which had a `get_config()` function before too. 
But is was not used, and it is used now after these commits. However, the `Mock()` object in the tests gave a `'Mock' object is not subsciptable` error when I wanted to access `model.optimizer.get_config()['name']`, where `model` is a `Mock` object. And to run the function, which resets the optimizer I need to get that attribute.

That is the reason for this line of code.

File: colearn_pytorch/test_pytorch_learner.py (hunk 6 above)
Comment (patbaa) [reply colearn_pytorch/test_pytorch_learner.py:30]: In Keras there is a `model.get_weight()`, which returns with all the weights (the trainable and the non-trainable too) as a list of numpy arrays.

In PyTorch there is a `model.parameters()` function, which returns all the trainable weights (and does not return the non-trainable at all, [see](https://discuss.pytorch.org/t/difference-between-state-dict-and-parameters/37531)) as a list of torch tensors.

Also in PyTorch there is a `model.state_dict()` function, which returns all the parameters (both trainable and non-trainable) as an OrderedDict. 

In the new update I am transitioning from `model.parameters()` to `model.state_dict()` to share the non-trainable parameters too (batch norm running mean, var for example). I changed the pytorch tests to use the state_dict() format. It does not affect the Keras learner.

File: colearn_keras/keras_learner.py (hunk 3 above)
Comment (qati) [reply colearn_keras/keras_learner.py]: cool

File: colearn_keras/test_keras_learner.py (hunk 4 above)
Comment (qati) [reply colearn_keras/test_keras_learner.py:31]: okay

File: colearn_pytorch/test_pytorch_learner.py (hunk 6 above)
Comment (qati) [reply colearn_pytorch/test_pytorch_learner.py:30]: so using state_dict is optional and it is still backward compatible?

Comment (patbaa) [reply colearn_pytorch/test_pytorch_learner.py:30]: both the model proposal and the model acceptance was modified. this way the demos are running without any modification, so it is backward compatible in that sense. (the exact model format is changed, that cannot be backward compatible, as we are saving/loading more information than previously, but the API part is unchanged)

**Review by evsmithx — APPROVED** [review]
